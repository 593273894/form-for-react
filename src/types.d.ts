declare interface FormInputProps {
    api: ContextApi;
    className?: string;
    name: string;
    label: string;
    value?: string | Array<string> | boolean | number;
    changeFieldValue: Function;
    disabled?: boolean;
}
declare interface ContextApi {
    getErrors: Function;
    setValue: Function;
    getValues: () => Object;
    getTouched: Function;
    addField: (field: Field) => void;
    changeField: (field: Field) => void;
    removeField: (name: string) => void;
}
declare interface FormContext {
    api: ContextApi;
}
declare interface Field {
    name: string;
    validate: () => null | string;
    required: boolean;
    requiredMessage?: string;
}