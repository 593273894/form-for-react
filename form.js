!function(e,t){"object"==typeof exports&&"object"==typeof module?module.exports=t(require("react"),require("classnames")):"function"==typeof define&&define.amd?define(["react","classnames"],t):"object"==typeof exports?exports["form-for-react"]=t(require("react"),require("classnames")):e["form-for-react"]=t(e.React,e.classnames)}(this,function(e,t){return function(e){function t(r){if(n[r])return n[r].exports;var o=n[r]={i:r,l:!1,exports:{}};return e[r].call(o.exports,o,o.exports,t),o.l=!0,o.exports}var n={};return t.m=e,t.c=n,t.i=function(e){return e},t.d=function(e,n,r){t.o(e,n)||Object.defineProperty(e,n,{configurable:!1,enumerable:!0,get:r})},t.n=function(e){var n=e&&e.__esModule?function(){return e.default}:function(){return e};return t.d(n,"a",n),n},t.o=function(e,t){return Object.prototype.hasOwnProperty.call(e,t)},t.p="",t(t.s=15)}([function(t,n){t.exports=e},function(e,t,n){"use strict";var r=n(0),o=n.n(r),i=n(5),a=n(4),s=n.n(a),u=n(2),c=n.n(u),l=n(3),p=this&&this.__extends||function(){var e=Object.setPrototypeOf||{__proto__:[]}instanceof Array&&function(e,t){e.__proto__=t}||function(e,t){for(var n in t)t.hasOwnProperty(n)&&(e[n]=t[n])};return function(t,n){function r(){this.constructor=t}e(t,n),t.prototype=null===n?Object.create(n):(r.prototype=n.prototype,new r)}}(),f=this&&this.__assign||Object.assign||function(e){for(var t,n=1,r=arguments.length;n<r;n++){t=arguments[n];for(var o in t)Object.prototype.hasOwnProperty.call(t,o)&&(e[o]=t[o])}return e},h=this&&this.__rest||function(e,t){var n={};for(var r in e)Object.prototype.hasOwnProperty.call(e,r)&&t.indexOf(r)<0&&(n[r]=e[r]);if(null!=e&&"function"==typeof Object.getOwnPropertySymbols)for(var o=0,r=Object.getOwnPropertySymbols(e);o<r.length;o++)t.indexOf(r[o])<0&&(n[r[o]]=e[r[o]]);return n},d=function(e){return t=function(t){function n(e){var n=t.call(this,e)||this;return n.propsValue=e.value,n.state={value:e.value},n}return p(n,t),n.prototype.componentDidMount=function(){this.context.api.addField({name:this.props.name,validate:this.props.validate,required:this.props.required,requiredMessage:this.props.requiredMessage||l.a}),void 0!==this.props.value&&""!==this.props.value&&this.context.api.setValue(this.props.name,this.props.value)},n.prototype.componentWillUnmount=function(){this.context.api.removeField(this.props.name)},n.prototype.componentWillReceiveProps=function(e){e.required!==this.props.required&&this.context.api.changeField({name:e.name,validate:e.validate,required:e.required}),this.isEqual(this.propsValue,e.value)||(this.propsValue=e.value,this.context.api.setValue(this.props.name,e.value),this.setState({value:e.value}))},n.prototype.isEqual=function(e,t){return"string"==typeof e||void 0===e||"boolean"==typeof e||"number"==typeof e?e===t:Array.isArray(e)?e.every(function(e){return t.indexOf(e)>=0})&&t.every(function(t){return e.indexOf(t)>=0}):(console.error("value should be a string, number, array or boolean"),!0)},n.prototype.renderLabel=function(){var e=this.props,t=e.label,n=e.required,r=e.name;return void 0!==t?o.a.createElement("label",{htmlFor:r,className:"field-label"},o.a.createElement("span",{className:"required"},n&&"*"),o.a.createElement("span",{className:"label"},t)):null},n.prototype.renderInput=function(){var t=this,n=this.props,r=n.name,i=n.onChange,a=(n.value,n.label,n.fieldClassName,n.fieldStyle,n.disabled,n.validate,n.requiredMessage,n.required,n.help,h(n,["name","onChange","value","label","fieldClassName","fieldStyle","disabled","validate","requiredMessage","required","help"])),s=this.context.api;return o.a.createElement("div",{className:"field-input"},o.a.createElement(e,f({api:this.context.api,name:r,value:this.state.value,changeFieldValue:function(e){i&&i(r,e),s.setValue(r,e),t.setState({value:e})}},a)))},n.prototype.renderHelp=function(){var e=this.props.help;return e&&o.a.createElement("span",{className:"field-help"},e)},n.prototype.renderError=function(){var e=this.props.name,t=this.context.api,n=t.getTouched();return(n=n[e]||n.__allTouched)&&o.a.createElement(i.a,{name:e})},n.prototype.render=function(){var e=this.props,t=e.fieldClassName,n=e.fieldStyle,r=e.disabled;return o.a.createElement("div",{className:s()("form-field",t,{disabled:r}),style:n},this.renderLabel(),this.renderInput(),this.renderHelp(),this.renderError())},n}(o.a.Component),t.contextTypes={api:c.a.object},t;var t};t.a=d},function(e,t,n){e.exports=n(13)()},function(e,t,n){"use strict";t.a="This field is required.";t.b=null},function(e,n){e.exports=t},function(e,t,n){"use strict";var r=n(0),o=n.n(r),i=n(2),a=n.n(i),s=n(4),u=n.n(s),c=n(3),l=this&&this.__extends||function(){var e=Object.setPrototypeOf||{__proto__:[]}instanceof Array&&function(e,t){e.__proto__=t}||function(e,t){for(var n in t)t.hasOwnProperty(n)&&(e[n]=t[n])};return function(t,n){function r(){this.constructor=t}e(t,n),t.prototype=null===n?Object.create(n):(r.prototype=n.prototype,new r)}}(),p=function(e){function t(t,n){var r=e.call(this,t,n)||this;return r.error=null,r}return l(t,e),t.prototype.shouldComponentUpdate=function(){var e=this.props.name,t=this.context.api,n=t.getErrors(e);return n!==this.error&&(this.error=n,!0)},t.prototype.render=function(){var e=this.props,t=e.name,n=e.className,r=this.context.api,i=r.getErrors()[t];return i!==c.b?o.a.createElement("span",{className:u()("validation-message",n)},i):null},t.contextTypes={api:a.a.object},t}(o.a.Component);t.a=p},function(e,t,n){"use strict";function r(){}var o=n(0),i=n.n(o),a=n(2),s=n.n(a),u=n(4),c=n.n(u),l=n(3),p=this&&this.__extends||function(){var e=Object.setPrototypeOf||{__proto__:[]}instanceof Array&&function(e,t){e.__proto__=t}||function(e,t){for(var n in t)t.hasOwnProperty(n)&&(e[n]=t[n])};return function(t,n){function r(){this.constructor=t}e(t,n),t.prototype=null===n?Object.create(n):(r.prototype=n.prototype,new r)}}(),f=this&&this.__assign||Object.assign||function(e){for(var t,n=1,r=arguments.length;n<r;n++){t=arguments[n];for(var o in t)Object.prototype.hasOwnProperty.call(t,o)&&(e[o]=t[o])}return e},h=this&&this.__rest||function(e,t){var n={};for(var r in e)Object.prototype.hasOwnProperty.call(e,r)&&t.indexOf(r)<0&&(n[r]=e[r]);if(null!=e&&"function"==typeof Object.getOwnPropertySymbols)for(var o=0,r=Object.getOwnPropertySymbols(e);o<r.length;o++)t.indexOf(r[o])<0&&(n[r[o]]=e[r[o]]);return n},d=function(e){function t(t){var n=e.call(this,t)||this;return n.state={errors:{}},n.values={},n.touched={__allTouched:!1},n.validates={},n.fields={},n}return p(t,e),t.prototype.getChildContext=function(){return{api:{getErrors:this.getErrors.bind(this),setValue:this.setValue.bind(this),getValues:this.getValues.bind(this),getTouched:this.getTouched.bind(this),addField:this.addField.bind(this),changeField:this.changeField.bind(this),removeField:this.removeField.bind(this)}}},t.prototype.addField=function(e){this.fields[e.name]=e,this.values[e.name]=""},t.prototype.changeField=function(e){this.fields[e.name]=e,this.setValue(e.name,this.values[e.name])},t.prototype.removeField=function(e){delete this.fields[e],this.values[e]=""},t.prototype.validateFields=function(){var e={};for(var t in this.fields){var n=this.fields[t];!function(e){return Array.isArray(e)?0===e.length:"number"==typeof e?void 0===e:!e}(this.values[t])?"function"==typeof this.fields[t].validate?e[t]=this.fields[t].validate(this.values[t],this.values):e[t]=l.b:n.required?e[t]=n.requiredMessage:e[t]=l.b}return e},t.prototype.setValue=function(e,t){this.values[e]=t,this.touched[e]=!0;var n=this.validateFields(),r=!0;for(var o in n)if(n[o]!==l.b){r=!1;break}this.setState({errors:n}),r?this.onValid():this.onInvalid()},t.prototype.getValues=function(){return this.values},t.prototype.getErrors=function(){return this.state.errors},t.prototype.getTouched=function(){return this.touched},t.prototype.onValid=function(){this.props.onValid&&this.props.onValid()},t.prototype.onInvalid=function(){this.props.onInvalid&&this.props.onInvalid()},t.prototype.onSubmit=function(e){e.preventDefault(),this.touched.__allTouched=!0;var t=this.validateFields(),n=!0;for(var r in t)if(t[r]!==l.b){n=!1;break}n&&this.props.onSubmit&&this.props.onSubmit(this.values),this.setState({errors:t})},t.prototype.render=function(){var e=this,t=this.props,n=t.className,r=(t.onInvalid,t.onSubmit,t.onValid,t.children,h(t,["className","onInvalid","onSubmit","onValid","children"]));return i.a.createElement("form",f({className:c()("form",n),onSubmit:function(t){return e.onSubmit(t)}},r),this.props.children)},t.defaultProps={onSubmit:r},t.childContextTypes={api:s.a.object},t}(i.a.Component);t.a=d},function(e,t,n){"use strict";var r=n(0),o=n.n(r),i=n(1),a=this&&this.__extends||function(){var e=Object.setPrototypeOf||{__proto__:[]}instanceof Array&&function(e,t){e.__proto__=t}||function(e,t){for(var n in t)t.hasOwnProperty(n)&&(e[n]=t[n])};return function(t,n){function r(){this.constructor=t}e(t,n),t.prototype=null===n?Object.create(n):(r.prototype=n.prototype,new r)}}(),s=[],u=function(e){function t(t){return e.call(this,t)||this}return a(t,e),t.prototype.render=function(){var e=this.props,t=e.value,n=void 0===t?s:t,r=e.options,i=e.changeFieldValue,a=e.name;return o.a.createElement("div",{className:"checkbox-group",id:a},r.map(function(e){var t=n.indexOf(e.value)>=0;return o.a.createElement("label",{key:e.value,className:"checkbox"+(t?" checked":"")},o.a.createElement("input",{name:e.name,className:"checkbox-input",type:"checkbox",checked:t,onChange:function(){i(t?n.filter(function(t){return t!==e.value}):n.concat([e.value]))}}),e.label)}))},t}(o.a.Component);t.a=n.i(i.a)(u)},function(e,t,n){"use strict";var r=n(0),o=n.n(r),i=n(1),a=this&&this.__extends||function(){var e=Object.setPrototypeOf||{__proto__:[]}instanceof Array&&function(e,t){e.__proto__=t}||function(e,t){for(var n in t)t.hasOwnProperty(n)&&(e[n]=t[n])};return function(t,n){function r(){this.constructor=t}e(t,n),t.prototype=null===n?Object.create(n):(r.prototype=n.prototype,new r)}}(),s=this&&this.__assign||Object.assign||function(e){for(var t,n=1,r=arguments.length;n<r;n++){t=arguments[n];for(var o in t)Object.prototype.hasOwnProperty.call(t,o)&&(e[o]=t[o])}return e},u=this&&this.__rest||function(e,t){var n={};for(var r in e)Object.prototype.hasOwnProperty.call(e,r)&&t.indexOf(r)<0&&(n[r]=e[r]);if(null!=e&&"function"==typeof Object.getOwnPropertySymbols)for(var o=0,r=Object.getOwnPropertySymbols(e);o<r.length;o++)t.indexOf(r[o])<0&&(n[r[o]]=e[r[o]]);return n},c=function(e){function t(){return null!==e&&e.apply(this,arguments)||this}return a(t,e),t.prototype.render=function(){var e=this.props,t=e.name,n=e.value,r=void 0===n?"":n,i=(e.api,e.changeFieldValue),a=u(e,["name","value","api","changeFieldValue"]);return o.a.createElement("input",s({id:t,className:"input",value:r,onChange:function(e){i(e.target.value)}},a))},t}(o.a.Component);t.a=n.i(i.a)(c)},function(e,t,n){"use strict";var r=n(0),o=n.n(r),i=n(1),a=this&&this.__extends||function(){var e=Object.setPrototypeOf||{__proto__:[]}instanceof Array&&function(e,t){e.__proto__=t}||function(e,t){for(var n in t)t.hasOwnProperty(n)&&(e[n]=t[n])};return function(t,n){function r(){this.constructor=t}e(t,n),t.prototype=null===n?Object.create(n):(r.prototype=n.prototype,new r)}}(),s=function(e){function t(t){return e.call(this,t)||this}return a(t,e),t.prototype.render=function(){var e=this.props,t=e.value,n=e.options,r=e.changeFieldValue,i=e.name;return o.a.createElement("div",{className:"radio-group",id:i},n.map(function(e){var n=e.value===t;return o.a.createElement("label",{key:e.value,className:"radio"+(n?" checked":"")},o.a.createElement("input",{name:e.name,className:"radio-input",type:"radio",checked:n,onChange:function(){r(e.value)}}),e.label)}))},t}(o.a.Component);t.a=n.i(i.a)(s)},function(e,t,n){"use strict";var r=n(0),o=n.n(r),i=n(1),a=this&&this.__extends||function(){var e=Object.setPrototypeOf||{__proto__:[]}instanceof Array&&function(e,t){e.__proto__=t}||function(e,t){for(var n in t)t.hasOwnProperty(n)&&(e[n]=t[n])};return function(t,n){function r(){this.constructor=t}e(t,n),t.prototype=null===n?Object.create(n):(r.prototype=n.prototype,new r)}}(),s=function(e){function t(t){return e.call(this,t)||this}return a(t,e),t.prototype.render=function(){var e=this.props,t=e.value,n=void 0===t?"":t,r=e.options,i=e.changeFieldValue,a=e.name;return o.a.createElement("select",{name:a,id:a,value:n,onChange:function(e){i(e.target.value)}},o.a.createElement("option",{value:""}," --- "),r.map(function(e){return o.a.createElement("option",{key:e.value,value:e.value},e.label)}))},t}(o.a.Component);t.a=n.i(i.a)(s)},function(e,t,n){"use strict";function r(e){return function(){return e}}var o=function(){};o.thatReturns=r,o.thatReturnsFalse=r(!1),o.thatReturnsTrue=r(!0),o.thatReturnsNull=r(null),o.thatReturnsThis=function(){return this},o.thatReturnsArgument=function(e){return e},e.exports=o},function(e,t,n){"use strict";function r(e,t,n,r,i,a,s,u){if(o(t),!e){var c;if(void 0===t)c=new Error("Minified exception occurred; use the non-minified dev environment for the full error message and additional helpful warnings.");else{var l=[n,r,i,a,s,u],p=0;c=new Error(t.replace(/%s/g,function(){return l[p++]})),c.name="Invariant Violation"}throw c.framesToPop=1,c}}var o=function(e){};e.exports=r},function(e,t,n){"use strict";var r=n(11),o=n(12),i=n(14);e.exports=function(){function e(e,t,n,r,a,s){s!==i&&o(!1,"Calling PropTypes validators directly is not supported by the `prop-types` package. Use PropTypes.checkPropTypes() to call them. Read more at http://fb.me/use-check-prop-types")}function t(){return e}e.isRequired=e;var n={array:e,bool:e,func:e,number:e,object:e,string:e,symbol:e,any:e,arrayOf:t,element:e,instanceOf:t,node:e,objectOf:t,oneOf:t,oneOfType:t,shape:t};return n.checkPropTypes=r,n.PropTypes=n,n}},function(e,t,n){"use strict";e.exports="SECRET_DO_NOT_PASS_THIS_OR_YOU_WILL_BE_FIRED"},function(e,t,n){"use strict";Object.defineProperty(t,"__esModule",{value:!0});var r=n(6),o=n(5),i=n(1),a=n(8),s=n(9),u=n(7),c=n(10);n.d(t,"Form",function(){return r.a}),n.d(t,"FormError",function(){return o.a}),n.d(t,"createFormField",function(){return i.a}),n.d(t,"Input",function(){return a.a}),n.d(t,"RadioGroup",function(){return s.a}),n.d(t,"CheckboxGroup",function(){return u.a}),n.d(t,"Select",function(){return c.a})}])});