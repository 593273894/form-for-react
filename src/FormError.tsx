import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';

export interface FormErrorProps {
    name: string;
    className?: string;
}

class FormError extends React.Component<FormErrorProps, any> {
    constructor(props, context) {
        super(props, context);
        this.error = null;
    }
    static contextTypes = {
        api: PropTypes.object
    }
    error: string | null;
    shouldComponentUpdate() {
        const { name } = this.props;
        const { api } = this.context;
        const error = api.getErrors(name);
        if (error !== this.error) {
            this.error = error;
            return true;
        }
        else {
            return false;
        }
    }
    render() {
        const { name, className } = this.props;
        const { api } = this.context;
        const error = api.getErrors()[name];
        return (
            error ? <span className={classnames('validation-message', className)}>{error}</span> : null
        );
    }
}

export default FormError;