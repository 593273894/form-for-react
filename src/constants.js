export const defaultRequiredMessage = 'this field is required';
export const NO_ERROR = null;