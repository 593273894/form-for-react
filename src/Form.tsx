import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import { NO_ERROR } from './constants';

function noop() { }
interface State {
    errors: Object;
}

export interface FormProps {
    onChange?: Function;
    onSubmit?: Function;
    onValid?: Function;
    onInvalid?: Function;
    className?: string;
}

export default class Form extends React.Component<FormProps, State> {
    constructor(props) {
        super(props);
        this.state = {
            errors: {}
        };
        this.values = {};
        this.touched = {
            __allTouched: false
        };
        this.validates = {};
        this.fields = {};
    }
    static defaultProps: FormProps = {
        onSubmit: noop,
    };
    static childContextTypes = {
        api: PropTypes.object
    };
    getChildContext() {
        return {
            api: {
                getErrors: this.getErrors.bind(this),
                setValue: this.setValue.bind(this),
                getValues: this.getValues.bind(this),
                getTouched: this.getTouched.bind(this),
                addField: this.addField.bind(this),
                changeField: this.changeField.bind(this),
                removeField: this.removeField.bind(this),
            }
        };
    }
    values: {};
    touched: { __allTouched: Boolean };
    validates: {};
    fields: {};
    addField(field) {
        this.fields[field.name] = field;
        this.values[field.name] = '';
    }
    changeField(field) {
        this.fields[field.name] = field;
        this.setValue(field.name, this.values[field.name]);
    }
    removeField(name) {
        delete this.fields[name];
        this.values[name] = '';
    }
    validateFields() {
        let errors = {};
        for (let name in this.fields) {
            const field = this.fields[name];
            if (isEmpty(this.values[name])) {
                if (field.required) {
                    errors[name] = field.requiredMessage;
                }
                else {
                    errors[name] = NO_ERROR;
                }
            }
            else {
                if (typeof this.fields[name].validate === 'function') {
                    errors[name] = this.fields[name].validate(this.values[name], this.values);
                }
                else {
                    errors[name] = NO_ERROR;
                }
            }
        }
        return errors;
        function isEmpty(value) {
            if (Array.isArray(value)) {
                return value.length === 0;
            }
            if (typeof value === 'number') {
                return value === undefined;
            }
            else {
                return !value;
            }
        }
    }
    setValue(name, value, triggerFormOnChange = false) {
        this.values[name] = value;
        this.touched[name] = true;
        const errors = this.validateFields();
        let noerror = true;
        for (let name in errors) {
            if (errors[name] !== NO_ERROR) {
                noerror = false;
                break;
            }
        }
        this.setState({
            errors
        });
        noerror ? this.onValid() : this.onInvalid();
        triggerFormOnChange && this.onchange(name, value);
    }
    getValues() {
        return this.values;
    }
    getErrors() {
        return this.state.errors;
    }
    getTouched() {
        return this.touched;
    }
    onchange(name, value) {
        this.props.onChange && this.props.onChange(name, value);
    }
    onValid() {
        this.props.onValid && this.props.onValid();
    }
    onInvalid() {
        this.props.onInvalid && this.props.onInvalid();
    }
    onSubmit(e) {
        e.preventDefault();
        this.touched.__allTouched = true;
        const errors = this.validateFields();
        let noerror = true;
        for (let name in errors) {
            if (errors[name] !== NO_ERROR) {
                noerror = false;
                break;
            }
        }
        if (noerror) {
            this.props.onSubmit && this.props.onSubmit(this.values);
        }
        this.setState({
            errors
        });
    }

    render() {
        const { className, onChange, onSubmit, onInvalid, onValid, children, ...rest } = this.props;
        return (
            <form className={classnames('form', className)} onSubmit={e => this.onSubmit(e)} {...rest} >
                {this.props.children}
            </form>
        );
    }
}