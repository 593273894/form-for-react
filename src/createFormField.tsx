import React from 'react';
import FormError from './FormError';
import classnames from 'classnames';
import PropTypes from 'prop-types';
import { defaultRequiredMessage } from './constants';
export interface FormFieldProps {
    name: string;
    label: string;
    validate: Function;
    onChange: (name: string, value: any) => any;
    required: boolean;
    value?: string | number;
    requiredMessage?: string;
    fieldClassName?: string;
    fieldStyle?: Object;
    disabled?: boolean;
    help?: string;
}

const createFromField = (Input) => {
    return class FormField extends React.Component<FormFieldProps, any> {
        constructor(props) {
            super(props);
            this.propsValue = props.value;
            this.state = {
                value: props.value
            };
        }
        propsValue: string;
        static contextTypes = {
            api: PropTypes.object
        };
        context: FormContext;
        componentDidMount() {
            this.context.api.addField({
                name: this.props.name,
                validate: this.props.validate,
                required: this.props.required,
                requiredMessage: this.props.requiredMessage || defaultRequiredMessage
            });
            if (this.props.value !== undefined && this.props.value !== '') {
                this.context.api.setValue(this.props.name, this.props.value);
            }
        }
        componentWillUnmount() {
            this.context.api.removeField(this.props.name);
        }
        componentWillReceiveProps(nextprops) {
            if (nextprops.required !== this.props.required) {
                this.context.api.changeField({
                    name: nextprops.name,
                    validate: nextprops.validate,
                    required: nextprops.required
                });
            }
            if (!this.isEqual(this.propsValue, nextprops.value)) {
                this.propsValue = nextprops.value;
                this.context.api.setValue(this.props.name, nextprops.value);
                this.setState({
                    value: nextprops.value
                });
            }
        }
        isEqual(value1, value2) {
            if (typeof value1 === 'string' || typeof value1 === 'undefined' || typeof value1 === 'boolean' ||
                typeof value1 === 'number') {
                return value1 === value2;
            }
            else if (Array.isArray(value1)) {
                return value1.every(item => value2.indexOf(item) >= 0) && value2.every(item => value1.indexOf(item) >= 0);
            }
            else {
                console.error('value should be a string, number, array or boolean');
                return true;
            }
        }
        renderLabel() {
            const { label, required, name } = this.props;
            if (label !== undefined) {
                return (
                    <label htmlFor={name} className="field-label">
                        <span className="required">{required && '*'}</span>
                        <span className="label">{label}</span>
                    </label>
                );
            }
            else {
                return null;
            }
        }
        renderInput() {
            const { name, label, onChange, value, disabled, ...rest } = this.props;
            const { api } = this.context;
            return (
                <div className="field-input">
                    <Input
                        api={this.context.api}
                        name={name} label={label}
                        value={this.state.value}
                        disabled={disabled}
                        changeFieldValue={value => {
                            onChange && onChange(name, value);
                            api.setValue(name, value);
                            this.setState({
                                value: value
                            });
                        }} {...rest} />
                </div>
            );
        }
        renderHelp() {
            const { help } = this.props;
            return (
                help && <span className="field-help">{help}</span>
            );
        }
        renderError() {
            const { name } = this.props;
            const { api } = this.context;
            let touched = api.getTouched();
            touched = touched[name] || touched.__allTouched;
            return (
                touched && <FormError name={name} />
            );
        }
        render() {
            const { fieldClassName, fieldStyle, disabled } = this.props;
            return (
                <div className={classnames('form-field', fieldClassName, { disabled })} style={fieldStyle} >
                    {this.renderLabel()}
                    {this.renderInput()}
                    {this.renderHelp()}
                    {this.renderError()}
                </div>
            );
        }
    };
};

export default createFromField;