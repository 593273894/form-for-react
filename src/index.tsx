import Form from './Form';
import FormError from './FormError';
import createFormField from './createFormField';

import Input from './components/Input';
import RadioGroup from './components/RadioGroup';
import CheckboxGroup from './components/CheckboxGroup';
import Select from './components/Select';

export {
    Form,
    FormError,
    createFormField,

    Input,
    RadioGroup,
    CheckboxGroup,
    Select
};