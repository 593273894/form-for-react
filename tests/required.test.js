import React from 'react';
import { Form, Input, RadioGroup, CheckboxGroup, Select } from './form';
import { defaultRequiredMessage } from '../src/constants';

test('show required message, and set a \'has-error\' className after touched', () => {
    const component = mount(
        <Form>
            <Input
                name="input"
                label="input"
                required />
        </Form>
    );
    expect(component).toMatchSnapshot();

    component.find('#input').simulate('change', { target: { value: '' } });
    let text = component.find('.form').childAt(0).find('.validation-message').text();
    expect(text).toEqual(defaultRequiredMessage);
    let hasError = component.find('.form').childAt(0).hasClass('has-error');
    expect(hasError).toEqual(true);

    expect(component).toMatchSnapshot();
});

test('show custom required message', () => {
    const requiredMessage = "please input this field";
    const component = mount(
        <Form>
            <Input
                name="input"
                label="input"
                required
                requiredMessage={requiredMessage} />
        </Form>
    );
    expect(component).toMatchSnapshot();

    component.find('#input').simulate('change', { target: { value: '' } });
    let text = component.find('.form').childAt(0).find('.validation-message').text();
    expect(text).toEqual(requiredMessage);

    let hasError = component.find('.form').childAt(0).hasClass('has-error');
    expect(hasError).toEqual(true);

    expect(component).toMatchSnapshot();
});

test('required message can be empty string', () => {
    const requiredMessage = "";
    const component = mount(
        <Form>
            <Input
                name="input"
                label="input"
                required
                requiredMessage={requiredMessage} />
        </Form>
    );
    expect(component).toMatchSnapshot();

    component.find('#input').simulate('change', { target: { value: '' } });
    let text = component.find('.form').childAt(0).find('.validation-message').text();
    expect(text).toEqual(requiredMessage);

    let hasError = component.find('.form').childAt(0).hasClass('has-error');
    expect(hasError).toEqual(true);

    expect(component).toMatchSnapshot();
});

test('show required message, and set a \'has-error\' className after form submit', () => {
    const component = mount(
        <Form>
            <Input
                name="input"
                label="input"
                required />
        </Form>
    );
    expect(component).toMatchSnapshot();

    component.find('.form').simulate('submit');
    let text = component.find('.form').childAt(0).find('.validation-message').text();
    expect(text).toEqual(defaultRequiredMessage);
    let hasError = component.find('.form').childAt(0).hasClass('has-error');
    expect(hasError).toEqual(true);

    expect(component).toMatchSnapshot();
});

