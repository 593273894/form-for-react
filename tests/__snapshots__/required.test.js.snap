// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`show custom required message 1`] = `
<Form
  onSubmit={[Function]}
>
  <form
    className="form"
    onSubmit={[Function]}
  >
    <FormField
      label="input"
      name="input"
      required={true}
      requiredMessage="please input this field"
    >
      <div
        className="form-field"
      >
        <label
          className="field-label"
          htmlFor="input"
        >
          <span
            className="required"
          >
            *
          </span>
          <span
            className="label"
          >
            input
          </span>
        </label>
        <div
          className="field-input"
        >
          <Input
            api={
              Object {
                "addField": [Function],
                "changeField": [Function],
                "getErrors": [Function],
                "getTouched": [Function],
                "getValues": [Function],
                "removeField": [Function],
                "setValue": [Function],
              }
            }
            changeFieldValue={[Function]}
            name="input"
          >
            <input
              className="input"
              id="input"
              onChange={[Function]}
              value=""
            />
          </Input>
        </div>
      </div>
    </FormField>
  </form>
</Form>
`;

exports[`show custom required message 2`] = `
<Form
  onSubmit={[Function]}
>
  <form
    className="form"
    onSubmit={[Function]}
  >
    <FormField
      label="input"
      name="input"
      required={true}
      requiredMessage="please input this field"
    >
      <div
        className="form-field"
      >
        <label
          className="field-label"
          htmlFor="input"
        >
          <span
            className="required"
          >
            *
          </span>
          <span
            className="label"
          >
            input
          </span>
        </label>
        <div
          className="field-input"
        >
          <Input
            api={
              Object {
                "addField": [Function],
                "changeField": [Function],
                "getErrors": [Function],
                "getTouched": [Function],
                "getValues": [Function],
                "removeField": [Function],
                "setValue": [Function],
              }
            }
            changeFieldValue={[Function]}
            name="input"
            value=""
          >
            <input
              className="input"
              id="input"
              onChange={[Function]}
              value=""
            />
          </Input>
        </div>
        <FormError
          name="input"
        >
          <span
            className="validation-message"
          >
            please input this field
          </span>
        </FormError>
      </div>
    </FormField>
  </form>
</Form>
`;

exports[`show required message after form submit 1`] = `
<Form
  onSubmit={[Function]}
>
  <form
    className="form"
    onSubmit={[Function]}
  >
    <FormField
      label="input"
      name="input"
      required={true}
    >
      <div
        className="form-field"
      >
        <label
          className="field-label"
          htmlFor="input"
        >
          <span
            className="required"
          >
            *
          </span>
          <span
            className="label"
          >
            input
          </span>
        </label>
        <div
          className="field-input"
        >
          <Input
            api={
              Object {
                "addField": [Function],
                "changeField": [Function],
                "getErrors": [Function],
                "getTouched": [Function],
                "getValues": [Function],
                "removeField": [Function],
                "setValue": [Function],
              }
            }
            changeFieldValue={[Function]}
            name="input"
          >
            <input
              className="input"
              id="input"
              onChange={[Function]}
              value=""
            />
          </Input>
        </div>
      </div>
    </FormField>
  </form>
</Form>
`;

exports[`show required message after form submit 2`] = `
<Form
  onSubmit={[Function]}
>
  <form
    className="form"
    onSubmit={[Function]}
  >
    <FormField
      label="input"
      name="input"
      required={true}
    >
      <div
        className="form-field"
      >
        <label
          className="field-label"
          htmlFor="input"
        >
          <span
            className="required"
          >
            *
          </span>
          <span
            className="label"
          >
            input
          </span>
        </label>
        <div
          className="field-input"
        >
          <Input
            api={
              Object {
                "addField": [Function],
                "changeField": [Function],
                "getErrors": [Function],
                "getTouched": [Function],
                "getValues": [Function],
                "removeField": [Function],
                "setValue": [Function],
              }
            }
            changeFieldValue={[Function]}
            name="input"
          >
            <input
              className="input"
              id="input"
              onChange={[Function]}
              value=""
            />
          </Input>
        </div>
        <FormError
          name="input"
        >
          <span
            className="validation-message"
          >
            This field is required.
          </span>
        </FormError>
      </div>
    </FormField>
  </form>
</Form>
`;

exports[`show required message after touched 1`] = `
<Form
  onSubmit={[Function]}
>
  <form
    className="form"
    onSubmit={[Function]}
  >
    <FormField
      label="input"
      name="input"
      required={true}
    >
      <div
        className="form-field"
      >
        <label
          className="field-label"
          htmlFor="input"
        >
          <span
            className="required"
          >
            *
          </span>
          <span
            className="label"
          >
            input
          </span>
        </label>
        <div
          className="field-input"
        >
          <Input
            api={
              Object {
                "addField": [Function],
                "changeField": [Function],
                "getErrors": [Function],
                "getTouched": [Function],
                "getValues": [Function],
                "removeField": [Function],
                "setValue": [Function],
              }
            }
            changeFieldValue={[Function]}
            name="input"
          >
            <input
              className="input"
              id="input"
              onChange={[Function]}
              value=""
            />
          </Input>
        </div>
      </div>
    </FormField>
  </form>
</Form>
`;

exports[`show required message after touched 2`] = `
<Form
  onSubmit={[Function]}
>
  <form
    className="form"
    onSubmit={[Function]}
  >
    <FormField
      label="input"
      name="input"
      required={true}
    >
      <div
        className="form-field"
      >
        <label
          className="field-label"
          htmlFor="input"
        >
          <span
            className="required"
          >
            *
          </span>
          <span
            className="label"
          >
            input
          </span>
        </label>
        <div
          className="field-input"
        >
          <Input
            api={
              Object {
                "addField": [Function],
                "changeField": [Function],
                "getErrors": [Function],
                "getTouched": [Function],
                "getValues": [Function],
                "removeField": [Function],
                "setValue": [Function],
              }
            }
            changeFieldValue={[Function]}
            name="input"
            value=""
          >
            <input
              className="input"
              id="input"
              onChange={[Function]}
              value=""
            />
          </Input>
        </div>
        <FormError
          name="input"
        >
          <span
            className="validation-message"
          >
            This field is required.
          </span>
        </FormError>
      </div>
    </FormField>
  </form>
</Form>
`;
